File: abc378b/abc378b.ml
open Printf
open Scanf

let gomi q r d = ((r - d) mod q + q) mod q + d

let () =
  scanf "%d " (fun n ->
      let tbl = Array.init n (fun _ -> scanf "%d %d " (fun q r -> gomi q r)) in
      scanf "%d " (fun q ->
          List.init q (fun _ -> scanf "%d %d " (fun t d -> tbl.(t - 1) d)) |> List.iter @@ printf "%d\n"))

File: abc378b/abc378b.hs
{-# LANGUAGE OverloadedStrings #-}
import Control.Monad (replicateM, forM_)
import Data.Vector (Vector, (!))
import qualified Data.Vector as V

main :: IO ()
main = do
  n <- readLn
  qrs <- replicateM n f
  let tbl = V.fromList $ map (\[q, r] -> gomi q r) qrs :: Vector (Int -> Int)
  q <- readLn
  tds <- replicateM q f
  forM_ tds $ \[t, d] -> do
    print $ solve tbl t d
  where
    f = map read <$> words <$> getLine

solve :: Vector (Int -> Int) -> Int -> Int -> Int
solve tbl t d = (tbl ! (t - 1)) d

gomi :: Int -> Int -> Int -> Int
gomi q r d = (r - d) `mod` q + d
